Sources/SwiftProtobuf/JSONScanner.swift

```diff
@@ -706,6 +398,325 @@ internal struct JSONScanner {
     return Character(UnicodeScalar(UInt32(currentByte))!)
   }
 
+  // Parse the leading UInt64 from the provided utf8 bytes.
+  //
+  // This is called in three different situations:
+  //
+  // * Unquoted number.
+  //
+  // * Simple quoted number.  If a number is quoted but has no
+  //   backslashes, the caller can use this directly on the UTF8 by
+  //   just verifying the quote marks.  This code returns `nil` if it
+  //   sees a backslash, in which case the caller will need to handle ...
+  //
+  // * Complex quoted number.  In this case, the caller must parse the
+  //   quoted value as a string, then convert the string to utf8 and
+  //   use this to parse the result.  This is slow but fortunately
+  //   rare.
+  //
+  // In the common case where the number is written in integer form,
+  // this code does a simple straight conversion.  If the number is in
+  // floating-point format, this uses a slower and less accurate
+  // approach: it identifies a substring comprising a float, and then
+  // uses Double() and UInt64() to convert that string to an unsigned
+  // integer.  In particular, it cannot preserve full 64-bit integer
+  // values when they are written in floating-point format.
+  //
+  // If it encounters a "\" backslash character, it returns a nil.  This
+  // is used by callers that are parsing quoted numbers.  See nextSInt()
+  // and nextUInt() below.
+  private func parseBareUInt64(
+    source: UnsafeBufferPointer<UInt8>,
+    index: inout UnsafeBufferPointer<UInt8>.Index,
+    end: UnsafeBufferPointer<UInt8>.Index
+  ) throws -> UInt64? {
+    if index == end {
+      throw JSONDecodingError.truncated
+    }
+    let start = index
+    let c = source[index]
+    switch c {
+    case asciiZero: // 0
+      source.formIndex(after: &index)
+      if index != end {
+        let after = source[index]
+        switch after {
+        case asciiZero...asciiNine: // 0...9
+          // leading '0' forbidden unless it is the only digit
+          throw JSONDecodingError.leadingZero
+        case asciiPeriod, asciiLowerE, asciiUpperE: // . e
+          // Slow path: JSON numbers can be written in floating-point notation
+          index = start
+          if let d = try parseBareDouble(source: source,
+                                         index: &index,
+                                         end: end) {
+            if let u = UInt64(exactly: d) {
+              return u
+            }
+          }
+          throw JSONDecodingError.malformedNumber
+        case asciiBackslash:
+          return nil
+        default:
+          return 0
+        }
+      }
+      return 0
+    case asciiOne...asciiNine: // 1...9
+      var n = 0 as UInt64
+      while index != end {
+        let digit = source[index]
+        switch digit {
+        case asciiZero...asciiNine: // 0...9
+          let val = UInt64(digit - asciiZero)
+          if (n > UInt64.max / 10
+              || (n == UInt64.max / 10 && val > UInt64.max % 10)) {
```

Comment:
```
(n > UInt64.max / 10) || (n * 10 > UInt64.max - value)
```
I like this form.  Especially since `UInt64.max / 10` is a constant and `n * 10` can be reused immediately below.  That's less total work than my current form.

I've also updated the same code in TextFormatScanner.